### PR TITLE
linker: warn about orphan sections

### PR DESCRIPTION
**Pull request by dcpleung**
This adds a linker flag and necessary changes to linker scripts so that linker will warn about orphan sections.

Relates to #5534

### CODE REVIEW

**Comment by dcpleung** [comment]
Sanitycheck passed. But marked as DNM for now so we can gauge the impact.

**Review by andyross — APPROVED** [review]
So... devil's advocacy: that is a LOT of change just to enable one warning.  And it's likely this is going to have to evolve along with us as e.g. the toolchain grows new hair (random thought: does LTO involve a bunch of new section names?  I bet it does).

I mean, it's a good warning.  But we're biting off a bunch of linker complexity to enable it.  Not sure the tradeoff is worth it personally.  But then again, you've already done all the work :)

**Comment by dcpleung** [comment]
There is an issue #9498 where the developer specifies a custom section which results in fault. The reason was because the linker "randomly" places the custom section in front of data/bss section at final linking. This moves the device structs around which invalidates the addresses being calculated using the prebuilt binary So the idea is to warn at build time (#5534). A lot of these sections (like the debug.* and rel.*) are included in the linker scripts by binutils to produce ELF binaries. And yes, new sections produced by compilers will now need to be specified manually.

File: include/arch/arm/cortex_m/scripts/linker.ld [hunk 1]
```diff
@@ -226,6 +270,19 @@ SECTIONS
 	} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
 #endif
 
+#ifdef CONFIG_SOC_SERIES_PSOC62
```
Comment (nashif) [inline include/arch/arm/cortex_m/scripts/linker.ld]: looks like we need a better way to add soc specific sections into this general .ld file. Adding SoC specific section will make this file explode with ifdefs.

Comment (dcpleung) [reply include/arch/arm/cortex_m/scripts/linker.ld]: Added something similar to the custom-*.ld for SoC.

File: CMakeLists.txt [hunk 2]
```diff
@@ -287,6 +287,13 @@ zephyr_ld_options(
   ${LINKERFLAGPREFIX},--build-id=none
   )
 
+# Funny thing is if you set this to =error, some architectures will
+# fail the compiler flag check and this will be removed from build.
+# So keep it at =warn.
```
Comment (SebastianBoe) [inline CMakeLists.txt]: Would

zephyr_link_libraries(
  ${LINKERFLAGPREFIX},--orphan-handling=error
  )

have the behaviour we want?

**Review by SebastianBoe — APPROVED** [review]
This improves the error feedback significantly for non-advanced users at the cost of added maintenance for advanced users, so this is very welcome.

File: soc/Kconfig [hunk 3]
```diff
@@ -11,3 +11,24 @@ osource "$(SOC_DIR)/$(ARCH)/Kconfig"
...
 
 endmenu
+
+#
+# SOC_*_LD: SoC specific Linker script additions
+#
+config SOC_NOINIT_LD
+	bool
+	help
+	  Include a SoC specific linker script fragment for inserting
```
Comment (SebastianBoe) [inline soc/Kconfig]: a SoC specific linker script fragment

an SoC specific linker script fragment named soc-noinit.ld

File: soc/Kconfig [hunk 4]
```diff
@@ -11,3 +11,24 @@ osource "$(SOC_DIR)/$(ARCH)/Kconfig"
 osource "$(SOC_DIR)/$(ARCH)/*/Kconfig"
 
 endmenu
+
+#
+# SOC_*_LD: SoC specific Linker script additions
```
Comment (SebastianBoe) [inline soc/Kconfig]: There is no reasonable way for a non-ARM user to discover that he can't use this feature. We either need to make the Kconfig options depend on ARM or add support for all platforms.

File: CMakeLists.txt (hunk 2 above)
Comment (dcpleung) [reply CMakeLists.txt]: The linker will fail if =error is used when there are orphan sections. However, the ARC and Xtensa, cmake somehow silently drops the flag. There is nothing to indicate problems in the CMakeError.log. Deliberately setting the wrong parameter (e.g. =badarg) would result in error output about invalid argument in CMakeError.log. I haven't been able to figure out why cmake would drop the flag. So we can only use =warn for now to enable warning for all architectures. Sanitycheck will still errors out if there are any orphan sections.

**Review by nashif — APPROVED** [review]

File: soc/Kconfig (hunk 4 above)
Comment (dcpleung) [reply soc/Kconfig]: Added the #include for common linker scripts under include/, and added conditionals for the Kconfig.

File: CMakeLists.txt (hunk 2 above)
Comment (SebastianBoe) [reply CMakeLists.txt]: I was able to reproduce your observation about the flag being dropped with zephyr_ld_options, but when using zephyr_link_libraries I got the desired behaviour. Error on orphan sections.

Note, using zephyr_link_libraries should be used in place of zephyr_ld_options.

zephyr_ld_options will by-design drop the flag if cmake is unable to compile a simple C file with it (presumably there are some orphan sections in the compiler flag test).

zephyr_link_libraries just adds the flag without any toolchain-capability test.

https://gist.github.com/SebastianBoe/d480b62d0f998d0f108ad75fba8ed1fc

zephyr-sdk-0.9.3
qemu_xtensa

**Comment by pfalcon** [comment]
@OlegKapshii: FYI, there were some psoc6 related changes in this PR (just so you know they were done).

**Comment by SebastianBoe** [comment]
For future reference, this PR may cause regressions.

We may discover that some of the input sections that we are now explicitly placing in output sections were previously placed in a different and accidentally "correct" position, or should actually be discarded instead of placed.

File: CMakeLists.txt (hunk 2 above)
Comment (dcpleung) [reply CMakeLists.txt]: I thought about it but I am hesitant on forcing this flag without compiler check. Putting this in zephyr_link_libraries assumes everyone will be using binutils to link the final binaries. As we support custom toolchain, that compiler/linker combo may not support this flag.

**Comment by pfalcon** [comment]
~~~
[100%] Linking C executable zephyr.elf
/usr/bin/ld: warning: orphan section `.rel._k_mutex.static.pthread_key_lock' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel._k_mem_slab.static.posix_timer_slab' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel._k_sem.static.pthread_key_sem' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel._k_sem.static.mq_sem' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
/usr/bin/ld: warning: orphan section `.rel.log_const_kernel' from `/usr/lib/gcc/x86_64-linux-gnu/7/../../../i386-linux-gnu/Scrt1.o' being placed in section `.rel.dyn'.
Generating files from zephyr.elf for board: native_posix
~~~

native_posix. @aescolar: FYI

**Comment by aescolar** [comment]
Thanks @pfalcon , but this would seem to be a bug in this PR that needs to be addressed by @dcpleung . @pfalcon could you create an issue an assign it to him?

**Comment by pfalcon** [comment]
> Thanks @pfalcon , but this would seem to be a bug in this PR

Well, maybe just misfeature, the resulting executables seems to work.

> @pfalcon could you create an issue an assign it to him?

Done: https://github.com/zephyrproject-rtos/zephyr/issues/10493

**Pull request merged**